### PR TITLE
Add CLion IDE developer documentation

### PR DESCRIPTION
**Pull request by JanVogelsang**
This PR adds developer documentation on how to setup the CLion IDE for NEST, which is a popular commercial IDE for C++ projects.

[edit] review docs [here](https://nest-simulator--3045.org.readthedocs.build/en/3045/developer_space/workflows/nest_with_ides.html#clion)

### CODE REVIEW

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst [hunk 1]
```diff
@@ -410,7 +410,42 @@ Code Completion in Xcode
...
+
+Setting up the project
+~~~~~~~~~~~~~~~~~~~~~~
+
+1. :ref:`Clone NEST <git_workflow>` onto your computer or select ``Get from VCS`` from CLion's start window
+2. First setup CMake
+
+.. image:: ../../static/img/clion_cmake_config.png
+
+3. Add the following CMake option: ``-DPython_ROOT_DIR=[path/to/root/of/python/env]``
```
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/workflows/nest_with_ides.rst:434]: Is this the only thing you need to add to the cmake config?

It is also unclear to me how you navigate from the cmake configuration to the run configuration based on the instructions and screenshots

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst [hunk 2]
```diff
@@ -410,7 +410,42 @@ Code Completion in Xcode
...
+4. Then add a run configuration
+
+.. image:: ../../static/img/clion_run_configuration.png
+
+5. Add a ``custom build application``
```
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/workflows/nest_with_ides.rst:439]: Again, unclear how you did this

**Review by jessica-mitchell — CHANGES_REQUESTED** [review]
thanks for adding this. However, I find the instructions not clear enough and I don't think all these screen shots are necessary. i would suggest being more explicit in the text and only keeping the run configuration screen shot and maybe the cmake one.
For example
" In the left panel select Build,Execution,Deployment -->CMake. Then add CMake options . . ."
" Navigate to the ??? to edit the run configuration.  "
Then explain which items need to be edited (e.g., "Change the working directory to the path to the source directory for NEST" ).

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst [hunk 3]
```diff
@@ -410,7 +410,42 @@ Code Completion in Xcode
...
-1. Remove all files and folders that are created with the new target.
-1. In the tab "Build Phase" of the new target, under "Compile Sources" add all ``*.h``, ``*.hpp``, ``*.c``, ``*.cc``, ``*.cpp`` files from the list (you can use CMD+a).
-1. Now Xcode generates its index and after that code completion should work.
+2. Make the target a OS X -> Command Line Tool (Next), of type C++ in your project (preselected). We call it ``completion``
+3. Remove all files and folders that are created with the new target.
+4. In the tab "Build Phase" of the new target, under "Compile Sources" add all ``*.h``, ``*.hpp``, ``*.c``, ``*.cc``, ``*.cpp`` files from the list (you can use CMD+a).
+5. Now Xcode generates its index and after that code completion should work.
+
+
+CLion
```
[the remark this answers is not in the document]
Comment (JanVogelsang) [reply doc/htmldoc/developer_space/workflows/nest_with_ides.rst:419]: You mean a link to the IDE product page? I.e., https://www.jetbrains.com/clion/ ?

**Comment by JanVogelsang** [comment]
Alright, I addressed your comments and made the instructions clearer. @otcathatsya Could you have a look at this as well and tell me if these instructions are clear enough?

**Comment by JanVogelsang** [comment]
It doesn't look very pretty yet, for some reason all lines that are followed by an unordered list are bold. I couldn't find out how to fix that, could you do that Jessica?

**Review by heplesser — CHANGES_REQUESTED** [review]
Some of the instructions are horribly outdated, e.g., mention OpenMPI 1.6 and MacPorts, which hardly anyone uses any more. I will try to take a look at the Mac part.

**Comment by otcathatsya** [comment]
> Alright, I addressed your comments and made the instructions clearer. @otcathatsya Could you have a look at this as well and tell me if these instructions are clear enough?

I think to a dev using CLion it's pretty clear, though maybe important to state that most of the steps described are to be able to properly debug NEST with the built-in IDE tools, since the "regular" project will work pretty much out of the box from my experience.

**Comment by JanVogelsang** [comment]
> Some of the instructions are horribly outdated, e.g., mention OpenMPI 1.6 and MacPorts, which hardly anyone uses any more. I will try to take a look at the Mac part.

Alright, please open a separate issue (or ideally also a PR) for that, as this is unrelated to the documentation for using CLion.

**Comment by JanVogelsang** [comment]
> > Alright, I addressed your comments and made the instructions clearer. @otcathatsya Could you have a look at this as well and tell me if these instructions are clear enough?
> 
> I think to a dev using CLion it's pretty clear, though maybe important to state that most of the steps described are to be able to properly debug NEST with the built-in IDE tools, since the "regular" project will work pretty much out of the box from my experience.

Right, that's true. The run configuration is required to debug C++ with CLion and the CMake configuration is fully optional, but helps with two things: 1. If not properly setup (e.g. just leaving the defaults), CLion reruns CMake from time to time (I think when checking out a new branch for example) and thus overwrites the files one generated when manually running CMake from the command line. 2. Even though I personally don't do that, one does not need the terminal at all anymore, as `cmake`, `make`, `make install`, and `python ..` are all run from CLion itself now.  
I'll add that to the documentation.

**Comment by terhorstd** [comment]
Note that I added a link to review the rendered documentation into the PR description.

**Comment by JanVogelsang** [comment]
> Note that I added a link to review the rendered documentation into the PR description.

Will this be automatically rebuild now that I made some more changes?

**Comment by jessica-mitchell** [comment]
@JanVogelsang FYI: here is the link to the Read the Docs build: https://nest-simulator--3045.org.readthedocs.build/en/3045/developer_space/workflows/nest_with_ides.html#clion

**Comment by jessica-mitchell** [comment]
> It doesn't look very pretty yet, for some reason all lines that are followed by an unordered list are bold. I couldn't find out how to fix that, could you do that Jessica?

This is because you need new lines around your second order bullet lists. Also in the first paragraph, you are missing a new line before the bullet list so the bullets are ignored.

Good rule of thumb in reStructured text is always include new lines around markup for headings, lists, directives etc.

**Comment by JanVogelsang** [comment]
Alright, thanks, it looks good now.

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst [hunk 4]
```diff
@@ -410,7 +410,48 @@ Code Completion in Xcode
...
+This recipe describes how to set up CLion for editing, building, and running NEST. The description here was tested on Ubuntu 23.  
+All the steps below are fully optional if you don't intend to debug your code using CLion and prefer to build NEST manually from the command line.  
+If you just want to debug your code, you can skip steps 1-2 and only follow steps 3-7. However, it is advised to also follow steps 1-2 to inform CLion of your desired CMake configuration.  
+Setting up the CMake configuration in CLion has the following advantages:  
+
+* CLion's default CMake configuration creates the build directory inside the source code root directory, which easily leads to unknowingly committing build files to git.
+* CLion regenerates build files when checking out a new branch, potentially overriding the build you created by manually running CMake from the command line.
+* After setting up both the CMake configuaration and a run configuration, you won't need to use the command line to run ``cmake``, ``make``, ``make install``, and ``python *.py`` at all anymore. All will be done automatically by CLion and you only need to run your application without having to manually recompile.
+
```
Comment (heplesser) [inline doc/htmldoc/developer_space/workflows/nest_with_ides.rst:430]: Please add this important bit of information:
```suggestion

.. note::

   CLion is a commercial product. It is *your responsibility* to ensure that you have a valid
   license permitting you to use CLion (or any software product) for your work on the    
   NEST Simulator. 
```

**Review by heplesser — CHANGES_REQUESTED** [review]
@JanVogelsang I see your point about not fixing outdated, unrelated material. I will look at that in a separate PR. Please see the suggestion I added on CLion.

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst (hunk 4 above)
Comment (JanVogelsang) [reply doc/htmldoc/developer_space/workflows/nest_with_ides.rst:430]: Agreed, this is indeed an important note for all commercial products!

**Review by heplesser — APPROVED** [review]

**Review by otcathatsya — APPROVED** [review]

**Comment by jessica-mitchell** [comment]
@JanVogelsang I wanted to create a PR against your branch because I had a couple of changes I was trying out myself. But your repo isn't listed in the base repositories. This happens quite frequently, and I thought it might be because of private repos but yours is public. Do you happen to know why this is?
These are the [changes](https://github.com/jessica-mitchell/nest-simulator/commit/c6b1d65af4a85a8985b85047656d83a9aa87ba71) I wanted to make. Could you apply the commit to your branch?

**Comment by JanVogelsang** [comment]
@jessica-mitchell No idea why that is the case, but sure, I can apply the commit.

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/workflows/nest_with_ides.rst:1]: I don't know how, but it seems like not all the changes that I did are present here.

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst [hunk 5]
```diff
@@ -409,7 +409,53 @@ Code Completion in Xcode
...
+   `CLion <https://www.jetbrains.com/clion/>`_ is a commercial product. It is *your responsibility* to ensure that you have a valid
+   license permitting you to use CLion (or any software product) for your work on the    
+   NEST Simulator. 
+
+Setting up the project
+~~~~~~~~~~~~~~~~~~~~~~
+
+1. :ref:`Clone NEST <git_workflow>` onto your computer or select ``Get from VCS`` from CLion's start window
+2. You will be prompted to setup CMake. You can find the CMake configuration in the settings (Build,Execution,Deployment --> CMake) to make more changes at a later point. Make the following changes and press "Ok":
```
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/workflows/nest_with_ides.rst:440]:
```suggestion
2. You will be prompted to setup CMake. You can find the CMake configuration in the settings (Go to :menuselection:`Build,Execution,Deployment --> CMake`) to make more changes at a later point. 

Make the following changes and press :menuselection:`Ok`:
```

**Review by jessica-mitchell — CHANGES_REQUESTED** [review]

File: doc/htmldoc/api_function_list.json
[the remark this answers is not in the document]
Comment (JanVogelsang) [reply doc/htmldoc/api_function_list.json:1]: Done

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst
Comment (JanVogelsang) [reply doc/htmldoc/developer_space/workflows/nest_with_ides.rst:1]: Which change are you missing? I thought I added everything..

File: doc/htmldoc/developer_space/workflows/nest_with_ides.rst (hunk 5 above)
Comment (JanVogelsang) [reply doc/htmldoc/developer_space/workflows/nest_with_ides.rst]: Done

**Pull request merged**